src/camera.c

```diff
@@ -0,0 +1,493 @@
+#include <nitro.h>
+#include <string.h>
+#include <nnsys.h>
+
+#include "data_02100844.h"
+
+#include "overlay115/camera_angle.h"
+
+#include "heap.h"
+#include "unk_0201E190.h"
+#include "camera.h"
+
+#define CAMERA_DEFAULT_ASPECT_RATIO     (FX32_ONE * 4 / 3)
+#define CAMERA_DEFAULT_NEAR_CLIP        (FX32_ONE * 150)
+#define CAMERA_DEFAULT_FAR_CLIP         (FX32_ONE * 900)
+
+GXBufferMode gBufferMode = GX_BUFFERMODE_W;
+
+static Camera *sActiveCamera = NULL;
+
+static void Camera_AdjustPositionAroundTarget(Camera *camera)
+{
+    u16 angleX = -camera->angle.x;
+    camera->lookAt.position.x = FX_Mul(FX_Mul(FX_SinIdx(camera->angle.y), camera->distance), FX_CosIdx(camera->angle.x));
+    camera->lookAt.position.z = FX_Mul(FX_Mul(FX_CosIdx(camera->angle.y), camera->distance), FX_CosIdx(camera->angle.x));
+    camera->lookAt.position.y = FX_Mul(FX_SinIdx(angleX), camera->distance);
+
+    VEC_Add(&camera->lookAt.position, &camera->lookAt.target, &camera->lookAt.position);
+}
+
+static void Camera_AdjustTargetFromAngle(Camera *camera)
+{
+    u16 angleX = -camera->angle.x;
+    camera->lookAt.target.x = -FX_Mul(FX_Mul(FX_SinIdx(camera->angle.y), camera->distance), FX_CosIdx(camera->angle.x));
+    camera->lookAt.target.z = -FX_Mul(FX_Mul(FX_CosIdx(camera->angle.y), camera->distance), FX_CosIdx(camera->angle.x));
+    camera->lookAt.target.y = -FX_Mul(FX_SinIdx(angleX), camera->distance);
+
+    VEC_Add(&camera->lookAt.target, &camera->lookAt.position, &camera->lookAt.target);
+}
+
+static void Camera_Init(const u16 fovY, Camera *camera)
+{
+    camera->fovY = fovY;
+    camera->perspective.sinFovY = FX_SinIdx(fovY);
+    camera->perspective.cosFovY = FX_CosIdx(fovY);
+    camera->perspective.aspectRatio = CAMERA_DEFAULT_ASPECT_RATIO;
+    camera->perspective.nearClip = CAMERA_DEFAULT_NEAR_CLIP;
+    camera->perspective.farClip = CAMERA_DEFAULT_FAR_CLIP;
+    camera->lookAt.up.x = 0;
+    camera->lookAt.up.y = FX32_ONE;
+    camera->lookAt.up.z = 0;
+    camera->targetPos = NULL;
+    camera->trackTargetX = FALSE;
+    camera->trackTargetY = FALSE;
+    camera->trackTargetZ = FALSE;
+    camera->history = NULL;
+}
+
+static void Camera_AdjustDeltaPos(Camera const *camera, VecFx32 *deltaPos)
+{
+    if (camera->trackTargetX == FALSE) {
+        deltaPos->x = 0;
+    }
+
+    if (camera->trackTargetY == FALSE) {
+        deltaPos->y = 0;
+    }
+
+    if (camera->trackTargetZ == FALSE) {
+        deltaPos->z = 0;
+    }
+}
+
+static void Camera_UpdateHistory(Camera const *camera, const VecFx32 *inPos, VecFx32 *outPos)
+{
+    if (camera->history == NULL) {
+        *outPos = *inPos;
+    } else {
```

Comment:
**suggestion:** Try using `camera->history == NULL` as a sentinel, e.g.:

```suggestion
    if (camera->history == NULL) {
        *outPos = *inPos;
        return;
    }
```

And then remove the `else` control.